### ISSUE TITLE
`all(x)` should remove `None` option from each element of `x`

### ISSUE THREAD

**Issue by mcrumiller** [astral-sh/ty#3543]
### Summary

When `all(x)` is called on an iterable, the elements in `x` cannot be `None`, but this is not narrowed:

```python
from typing import reveal_type

x = [1, 2, 3, None, 4, 5]

if all(x):
    for value in x:
        reveal_type(value)  # int | None, should be int
```

### Version

0.0.34

**Comment by carljm** [comment]
Thanks! It's a little tricky to get this right, which is I suspect why no Python type checker currently implements this narrowing. A list is mutable, so any addition to the list after `if all(x)` could render the narrowing invalid. If we actually narrow the type of the list from `list[int | None]` to `list[int]`, that should prevent unsafe additions to the list within the narrowed region, but is also wrong due to invariance of list: `list[int]` is not a subtype of `list[int | None]`, so we'd actually be narrowing to an incompatible type, which will cause other problems.

That said, there are other forms of common narrowing that are technically unsound but practically useful, and I suspect this could also be added. Won't be a short-term priority, though.
